### PR TITLE
Add a way for checkers to bias type arg inference toward defaults (Fixes #2686)

### LINKED ISSUE #2686: Type inference is too eager to use Constant Value Checker types
Consider the following code.

```
import org.checkerframework.common.value.qual.UnknownVal;

public class MyTree<Value> {

    public static <V> MyTree<V> newTree(V value) {
        throw new Error("body doesn't matter");
    }

    public MyTree<Value> put(String newKey, Value newValue) {
        throw new Error("body doesn't matter");
    }

    public void client1() {
        // type-checking error
        MyTree<String> root2 = MyTree.newTree("constructorarg").put("key1", "value1");
    }

    public void client2() {
        // OK
        MyTree<String> root2 = MyTree.newTree((@UnknownVal String) "constructorarg").put("key1", "value1");
    }

    public void client3() {
        // OK
        MyTree<String> root2 = MyTree.<String>newTree((@UnknownVal String) "constructorarg").put("key1", "value1");
    }

}
```

You can typecheck it for the Constant Value checker by running

	${CHECKERFRAMEWORK}/checker/bin/javac -processor org.checkerframework.common.value.ValueChecker MyTree.java

Type-checking of client1 fails with the messages:

```
MyTree.java:15: error: [assignment.type.incompatible] incompatible types in assignment.
        MyTree<String> root2 = MyTree.newTree("constructorarg").put("key1", "value1");
                                                                   ^
  found   : @UnknownVal MyTree<@StringVal("constructorarg") String>
  required: @UnknownVal MyTree<@UnknownVal String>
MyTree.java:15: error: [argument.type.incompatible] incompatible types in argument.
        MyTree<String> root2 = MyTree.newTree("constructorarg").put("key1", "value1");
                                                                            ^
  found   : @StringVal("value1") String
  required: @StringVal("constructorarg") String
```

The problem is that in the call `MyTree.newTree("constructorarg")`, the Checker Framework infers a constant-value type for "constructorarg", and then instantiates the constructor with that constant-value type.

Methods `client2` and `client3` show how to solve the problem, but both of them require code changes to the user code.  Is there a way to solve the problem without code changes to the user code?  It would be acceptable to not use constant-value types anywhere that type inference occurs.

### PR DESCRIPTION
**Pull request by smillst**
Currently only the Value Checker uses this.
Fixes #2686.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -2258,4 +2258,15 @@ public int getMinLenFromString(String sequenceExpression, Tree tree, TreePath cu
 
         return getMinLenValue(lengthAnno);
     }
+
+    @Override
+    public AnnotatedTypeMirror getDummyAssignedTo(ExpressionTree expressionTree) {
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: A comment would be helpful.  This returns the default annotations.

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 2]
```diff
@@ -3993,4 +3993,21 @@ public static String negateConstant(String constantExpression) {
             return "-" + constantExpression;
         }
     }
+
+    /**
+     * Creates an annotated type mirror that type argument inference should pretend the {@code
+     * expression} is assigned to. This implementation returns null, but subclasses may override
+     * this method to return a type.
+     *
+     * <p>This biases the inference algorithm toward default annotations for type variables in
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: I think that "this" refers to "override this method" (but not to this implementation).  One choice for overriding is to return default annotations, which biases as this sentence does, but other choices are possible.

The documentation doesn't say what null means or what the effect of returning null is.

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 3]
```diff
@@ -3993,4 +3993,21 @@ public static String negateConstant(String constantExpression) {
             return "-" + constantExpression;
         }
     }
+
+    /**
+     * Creates an annotated type mirror that type argument inference should pretend the {@code
+     * expression} is assigned to. This implementation returns null, but subclasses may override
+     * this method to return a type.
+     *
+     * <p>This biases the inference algorithm toward the annotations for type variables in invariant
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: I'm still not sure whether "This" refers to "This implementation returns null" or "subclasses may override this method to return a type."  Can you clarify?  I would put "This implementation returns null" and "subclasses may override this method to return a type" in different paragraphs, with "This biases..." in one of those paragraphs.  Thanks!

Comment (smillst) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: I wasn't done addressing your review.  (That's why I didn't reassign you.)  I was just pushing changes to get them to the remote branch.

Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: OK.  Sorry about that.

Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: (I just had a moment between other tasks...)

**Comment by smillst** [comment]
@mernst, before we merge this code, could you test that it solves the issue in original code?

**Comment by mernst** [comment]
> @mernst, before we merge this code, could you test that it solves the issue in original code?

I'm inclined to merge, because there is a test case and so this resolves the issue.  If there are still problems in the original code, then they probably deserve a different issue number.

**Pull request merged**